lsbatch/man1/bqueues.1
.ds ]W %
.ds ]L
.nh
.TH bqueues 1 "OpenLava Version 3.3 - Mar 2016"
.br
.SH NAME
\fBbqueues\fR - displays information about queues
.SH SYNOPSIS
.BR
.PP
.PP
\fBbqueues \fR[\fB-w\fR | \fB-l\fR] [\fB-m\fR \fIhost_name\fR\fI \fR|\fI \fR\fB-m\fR\fI host_group\fR\fI | \fR\fB-m all\fR]
[\fB-u\fR\fB \fR\fIuser_name \fR|\fI \fR\fB-u\fR\fI user_group | \fR\fB-u all\fR] [\fIqueue_name\fR ...]
.PP
\fBbqueues \fR[\fB-h\fR | \fB-V\fR]
.SH DESCRIPTION
.BR
.PP
.PP
\fB\fRDisplays information about queues.
.PP
By default, returns the following information about all queues: queue
name, queue priority, queue status, job slot statistics, and job state
statistics.
.PP
Batch queue names and characteristics are set up by the openlava
administrator (see lsb.queues(5) and mbatchd(8)).
.SH OPTIONS
.BR
.PP
.TP
\fB-w\fR

.IP
Displays queue information in a wide format. Fields are displayed
without truncation.


.TP
\fB-l
\fR
.IP
Displays queue information in a long multi-line format. The -l option
displays the following additional information: queue description,
queue characteristics and statistics, scheduling parameters, resource
limits, scheduling policies, users, hosts, user shares, windows,
associated commands, and job controls.


.TP
\fB-m\fR \fIhost_name \fR|\fI \fR\fB-m\fR\fI host_group \fR| \fB-m\fR\fI \fR\fBall\fR

.IP
Displays the queues that can run jobs on the specified host or host
group. If the keyword all is specified, displays the queues that can run
jobs on all hosts . For a list of host groups see bmgroup(1).


.TP
\fB-u\fR \fIuser_name \fR|\fI \fR\fB-u\fR\fI user_group\fR\fB \fR|\fB -u all\fR

.IP
Displays the queues that can accept jobs from the specified user or user
group (For a list of user groups see bugroup(1).) If the keyword `all'
is specified, displays the queues that can accept jobs from all users.


.TP
\fIqueue_name \fR...

.IP
Displays information about the specified queues.


.TP
\fB-h\fR

.IP
Prints command usage to stderr and exits.


.TP
\fB-V
\fR
.IP
Prints openlava release version to stdout and exits.


.SH OUTPUT
.BR
.PP
.SS Default Output
.BR
.PP
.PP
Displays the following fields:

.IP
QUEUE_NAME
.BR
.RS
.IP
The name of the queue. Queues are named to correspond to the
type of jobs usually submitted to them, or to the type of services
they provide.


.IP
lost_and_found
.BR
.RS
.IP
If the openlava administrator removes queues from the system, openlava
creates a queue called lost_and_found and places the jobs
from the removed queues into the lost_and_found queue.
Jobs in the lost_and_found queue will not be started unless
they are switched to other queues (see bswitch).

.RE
.RE
.RE

.IP
PARAMETER/STATISTICS
.BR
.RS

.IP
PRIO
.BR
.RS
.IP
The priority of the queue. If job priority is not configured,
determines the queue search order at job dispatch, suspension
and resumption time. Queues with higher priority values are
searched first for job dispatch and resumption (this is contrary
to UNIX process priority ordering), and queues with higher
priority values are searched last for job suspension.

.RE

.IP
STATUS
.BR
.RS
.IP
The current status of the queue. The possible values are:


.IP
Open
.BR
.RS
.IP
The queue is able to accept jobs.

.RE

.IP
Closed
.BR
.RS
.IP
The queue is not able to accept jobs.

.RE

.IP
Active
.BR
.RS
.IP
Jobs in the queue may be started.

.RE

.IP
Inactive
.BR
.RS
.IP
Jobs in the queue cannot be started for the time being.

.RE
.IP
At any moment, each queue is either Open or Closed, and is
either Active or Inactive. The queue can be opened, closed,
inactivated and re-activated by the openlava administrator using
badmin (see badmin(8)). The queue becomes inactive when
either its dispatch window is closed or its run window is closed
(see DISPATCH_WINDOWS in the "Output for the -l Option"
section). In this case, the queue cannot be activated using
badmin. The queue is re-activated by openlava when one of its
dispatch windows and one of its run windows are open again.
The initial state of a queue at openlava boot time is set to open, and
either active or inactive depending on its windows.

.RE
.RE

.IP
MAX
.BR
.RS
.IP
The maximum number of job slots that can be used by the jobs
from the queue. These job slots are used by dispatched jobs
which have not yet finished, and by pending jobs which have
slots reserved for them. A sequential job will use one job slot
when it is dispatched to a host, while a parallel job will use as
many job slots as is required by bsub -n when it is dispatched.
See bsub(1) for details. If `-' is displayed, there is no limit.

.RE

.IP
JL/U
.BR
.RS
.IP
The maximum number of job slots you can use for your jobs in
the queue. These job slots are used by your dispatched jobs
which have not yet finished, and by pending jobs which have
slots reserved for them. If `-' is displayed, there is no limit.

.RE

.IP
JL/P
.BR
.RS
.IP
The maximum number of job slots a processor can process
from the queue. This includes job slots of dispatched jobs that
have not yet finished, and job slots reserved for some pending
jobs. The job slot limit per processor (JL/P) controls the number
of jobs sent to each host. This limit is configured per processor
so that multiprocessor hosts are automatically allowed to run
more jobs. If `-' is displayed, there is no limit.

.RE

.IP
JL/H
.BR
.RS
.IP
The maximum number of job slots a host can process from the
queue. This includes the job slots of dispatched jobs that have
not yet finished, and those reserved for some pending jobs. The
job slot limit per host (JL/H) controls the number of jobs sent
to each host, regardless of whether a host is a uniprocessor host
or a multiprocessor host. If `-' is displayed, there is no limit.

.RE

.IP
NJOBS
.BR
.RS
.IP
The total number of job slots held currently by jobs in the
queue. This includes pending, running, suspended and
reserved job slots. A parallel job that is running on \fIn\fR processors
is counted as \fIn\fR job slots, since it takes \fIn\fR job slots in the queue.
See bjobs(1) for an explanation of batch job states.

.RE

.IP
PEND
.BR
.RS
.IP
The number of pending job slots in the queue.

.RE

.IP
RUN
.BR
.RS
.IP
The number of running job slots in the queue.

.RE

.IP
SUSP
.BR
.RS
.IP
The number of suspended job slots in the queue.

.RE
.SS Output for \fB-\fRl Option
.BR
.PP
.PP
In addition to the above fields, the \fB-\fRl option displays the following:

.IP
Description
.BR
.RS
.IP
A description of the typical use of the queue.

.RE

.IP
PARAMETERS/STATISTICS
.BR
.RS

.IP
NICE
.BR
.RS
.IP
The nice value at which jobs in the queue will be run. This is
the UNIX nice value for reducing the process priority (see
nice(1)).

.RE

.IP
STATUS
.BR
.RS

.IP
Inactive
.BR
.RS
.IP
The long format for the \fB-\fRl option gives the possible reasons
for a queue to be inactive:

.RE

.IP
Inact_Win
.BR
.RS
.IP
The queue is out of its dispatch window or its run window.

.RE

.IP
Inact_Adm
.BR
.RS
.IP
The queue has been inactivated by the openlava administrator.

.RE

.IP
SSUSP
.BR
.RS
.IP
The number of job slots in the queue allocated to jobs that are
suspended by openlava.

.RE

.IP
USUSP
.BR
.RS
.IP
The number of job slots in the queue allocated to jobs that are
suspended by the job submitter or by the openlava administrator.

.RE

.IP
RSV
.BR
.RS
.IP
The numbers of job slots in the queue that are reserved by openlava
for pending jobs.

.RE

.IP
Migration threshold
.BR
.RS
.IP
\fB\fRThe length of time in seconds that a job dispatched from the queue
will remain suspended by the system before openlava attempts to
migrate the job to another host. See the MIG parameter in
lsb.queues and lsb.hosts.

.RE

.IP
Schedule delay for a new job
.BR
.RS
.IP
The delay time in seconds for scheduling a session after a new job
is submitted. If the schedule delay time is zero, a new scheduling
session is started as soon as the job is submitted to the queue. See
the NEW_JOB_SCHEDULE_DELAY parameter in lsb.queues.

.RE

.IP
Interval for a host to accept two jobs
.BR
.RS
.IP
The length of time in seconds to wait after dispatching a job to a
host before dispatching a second job to the same host. If the job
accept interval is zero, a host may accept more than one job in each
dispatching interval. See the JOB_ACCEPT_INTERVAL parameter in
lsb.queues and lsb.params.

.RE

.IP
RESOURCE LIMITS
.BR
.RS
.IP

NEWS
Release notes for OpenLava versions

3.3.pre.2
=========
-- Keep the lsb.events and lsb.acct files always open to improve
   file access performance.
-- Implement the switching of the lsb.events asynchronously in mbatchd
   child process.
-- Implemented the bjdep command. bjdep jobid prints the status of
   job dependencies for the given job.
-- Add the bjdep man page.
-- Implement NIOS_RWAIT_SELECT, if configure in lsf.conf then nios
   and ls_rwait() will not use the SIGUSR1 to detect a task that
   has finished. ls_rwait() will select() on a connection with nios
   instead and nios will send a message when a task exits.
-- Fix RES NIOS protocol bug that prevented multiple tasks running on
   the same host from being collected using ls_rwait().
-- Implemented job groups. New commands are bgadd, bgdel and bgroups.
-- Implemented bjobs -UF unformatted bjobs output.
-- Add LSF_BINDIR in openlava startup scripts.
-- Implement bjobs -g to display only selected job groups.
-- Implement bkill -g to kill only jobs in a specific job group.

3.3.pre.1
=========
-- Fix fairshare algorithm to decrease the number of jobs per user
   only if the job has been started.
-- Fix a bug in ownership policy where wrong runtime comparison was
   performed.
-- Implement the replay of job absolute runtime for ownership policy.
-- Implement floating client feature. If the user configures in lsf.conf
   LIM_ACCEPT_FLOAT_CLIENT=y any host can submit jobs and query the
   daemons without being registered in the lsf.cluster file.

3.2.1
======
-- Fix the parsing algorithm for compact host name.

3.2.0
======
-- Fix main Makefile.am to include the m4 directory.

3.2.beta1
=========
-- Hosts member od a host group can now be specified in a compact format
   for example: (ol[1-3]).
-- Implement configure option --enable-debug. If set the compilation will
   have the -Werror option enabled. By default this option is not set.

3.2.pre
========
-- badmin ckconfig/reconfig displays configuration error and warning
   messages.
-- bsub/bmod/brestart -h displays usage information.
-- Added USER_SHARES information in lsb.users.5
-- Add lsclusters command.
-- Use qsort() to sort jobs in MBD this improves the performance
   of replay and general bsub operations.
-- Introduce a getc2() functions to read lsb.events in memory
   before parsing it.
-- Reduce the number of malloc operations when reading the
   lsb.events file.
-- Implement memory reservation if the queue has SLOT_MEMORY_RESERVE configured
   and the job requests rusage of memory. The jobs has to reserve slots
   on the hosts as well.
-- Implement a correct switching of the lim.events file.
-- Update the bmgroup command to display GROUP_SLOT and MAX_SLOTS
   sections if configured.
-- Update the bqueues command to display slot and memory reservation.
-- Allow batch hosts configured in lsb.hosts to be specified in a
   compact form, i.e. ol[1-3] the mbatchd will expand example this into 3
   hosts ol1, ol2, ol3.
-- Allow OpenLava to use cores as cpus regardless is the machines has
   hyperthreads or not. To enable this behaviour set the variable
   LIM_DEFINE_NCPUS=cores|threads in lsf.conf. Default is threads.
-- Fix bsub bug in which if user specify -n min,max the max was ignored.
-- Make fairshare configuration compatible with the LSF one.

3.1.4
======

-- Fix the problem that if the order of events in lim.events is not correct
   LIM may core dump.
-- Change the ownership of the mbatchd log file to avoid messages going
   to syslog messages file.
-- Prevent LIM from core dumping in findHost() if the input name is NULL.
-- Enable the secondary admin to run 'badmin reconfig'.
-- Fix res nios protocol to pass the correct taskid to nios.

3.1.3
======
-- Corrected the lsf.cluster man page describing LIM communication
   parameters.
-- Enable LIM communication debugging using LC_COMM class and
   LOG_INFO level.
-- Fix bug in announceMaster() which was slowing down announcement
   of slave hosts.
-- Fix a bug in the slave host timer causing slave to send load
   to master more frequently then necessary.
-- Fix counter update in fairshare policy when jobs are suspended

lsbatch/cmd/bqueues.c
/*
 * Copyright (C) 2015 - 2016 David Bigagli
 * Copyright (C) 2007 Platform Computing Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA
 *
 */

#include "cmd.h"
#include "../../lsf/intlib/sshare.h"

static void prtQueuesLong(int, struct queueInfoEnt *);
static void prtQueuesShort(int, struct queueInfoEnt *);
static void print_slot_shares(struct queueInfoEnt *);
static void print_slot_owned(struct queueInfoEnt *);

static char wflag = FALSE;
extern int terminateWhen_(int *, char *);

#define QUEUE_NAME_LENGTH    15
#define QUEUE_PRIO_LENGTH    4
#define QUEUE_STATUS_LENGTH  14
#define QUEUE_JL_U_LENGTH    4
#define QUEUE_JL_P_LENGTH    4
#define QUEUE_JL_H_LENGTH    4
#define QUEUE_MAX_LENGTH     4
#define QUEUE_NJOBS_LENGTH   5
#define QUEUE_PEND_LENGTH    5
#define QUEUE_RUN_LENGTH     5
#define QUEUE_SUSP_LENGTH    5
#define QUEUE_SSUSP_LENGTH   5
#define QUEUE_USUSP_LENGTH   5
#define QUEUE_NICE_LENGTH    4
#define QUEUE_RSV_LENGTH     4

void
usage(const char *cmd)
{
    fprintf(stderr, ": %s [-h] [-V] [-w | -l] [-m host_name | -m cluster_name]\n", cmd);
    fprintf(stderr, " [-u user_name]");
    fprintf(stderr, " [queue_name ...]\n");
}

int
main(int argc, char **argv)
{
    int numQueues;
    char **queues;
    struct queueInfoEnt *queueInfo;
    char lflag = FALSE;
    int cc;
    char *host;
    char *user;

    numQueues = 0;
    user = host = NULL;

    if (lsb_init(argv[0]) < 0) {
        lsb_perror("lsb_init");
        return -1;
    }

    while ((cc = getopt(argc, argv, "Vhlwm:u:")) != EOF) {
        switch (cc) {
            case 'l':
                lflag = TRUE;
                if (wflag) {
                    usage(argv[0]);
                    return -1;
                }
                break;
            case 'w':
                wflag = TRUE;
                if (lflag) {
                    usage(argv[0]);
                    return -1;
                }
                break;
            case 'm':
                if (host != NULL || *optarg == '\0')
                    usage(argv[0]);
                host = optarg;
                break;
            case 'u':
                if (user != NULL || *optarg == '\0')
                    usage(argv[0]);
                user = optarg;
                break;
            case 'V':
                fputs(_LS_VERSION_, stderr);
                return 0;
            case 'h':
            default:
                usage(argv[0]);
            return -1;
        }
    }

    queues = NULL;
    numQueues = 0;

    if (optind < argc) {
        numQueues = argc - optind;
        queues = calloc(argc - optind, sizeof(char *));
        for (cc = 0; cc < argc - optind; cc++)
            queues[cc] = argv[optind + cc];
    }

    TIMEIT(0, (queueInfo = lsb_queueinfo(queues,
                                         &numQueues,
                                         host,
                                         user,
                                         0)), "lsb_queueinfo");

    if (!queueInfo) {
        if (lsberrno == LSBE_BAD_QUEUE && queues)
            lsb_perror(queues[numQueues]);
        else {
            switch (lsberrno) {
                case LSBE_BAD_HOST   :
                case LSBE_QUEUE_HOST :
                    lsb_perror(host);
                    break;
                case LSBE_BAD_USER   :
                case LSBE_QUEUE_USE  :
                    lsb_perror(user);
                    break;
                default :
                    lsb_perror(NULL);
            }
        }
        _free_(queues);
        return -1;
    }

    if (lflag)
        prtQueuesLong(numQueues, queueInfo);
    else
        prtQueuesShort(numQueues, queueInfo);

    _free_(queues);

    return 0;
}

/* prtQueuesLong()
 */
static void
prtQueuesLong(int numQueues, struct queueInfoEnt *queueInfo)
{
    struct queueInfoEnt *qp;
    char *status;
    char userJobLimit[MAX_CHARLEN];
    char procJobLimit[MAX_CHARLEN];
    char hostJobLimit[MAX_CHARLEN];
    char maxJobs[MAX_CHARLEN];
    int i;
    int numDefaults = 0;
    struct lsInfo *lsInfo;
    int printFlag = 0;
    int printFlag1 = 0;
    int printFlag2 = 0;
    int printFlag3 = 0;
    int procLimits[3];

    if ((lsInfo = ls_info()) == NULL) {
        ls_perror("ls_info");
        exit(-1);
    }

    printf("\n");

    for (i = 0; i < numQueues; i++ )  {
        qp = &(queueInfo[i]);
        if (qp->qAttrib & Q_ATTRIB_DEFAULT)
            numDefaults++;
    }

    for (i = 0; i < numQueues; i++ ) {

        qp = &(queueInfo[i]);

        if (qp->qStatus & QUEUE_STAT_OPEN) {
            if ((qp->qStatus & QUEUE_STAT_ACTIVE)
                && (qp->qStatus & QUEUE_STAT_RUN))
                status = "   Open:Active  ";
            if ((qp->qStatus & QUEUE_STAT_ACTIVE)
                && !(qp->qStatus & QUEUE_STAT_RUN))
                status = " Open:Inact_Win "; /* windows are closed */
            if (!(qp->qStatus & QUEUE_STAT_ACTIVE))
                status = " Open:Inact_Adm "; /* By user (manager) */
        } else {  /* queue is disabled */
            if ((qp->qStatus & QUEUE_STAT_ACTIVE)
                && (qp->qStatus & QUEUE_STAT_RUN))
                status = "  Closed:Active ";
            if ((qp->qStatus & QUEUE_STAT_ACTIVE)
                && !(qp->qStatus & QUEUE_STAT_RUN))
                status = "Closed:Inact_Win"; /* by time window */
            if (!(qp->qStatus & QUEUE_STAT_ACTIVE))
                status = "Closed:Inact_Adm"; /* by user (manager) */
        }

        if (qp->maxJobs < INFINIT_INT)
            sprintf (maxJobs, "%5d", qp->maxJobs);
        else
            sprintf (maxJobs, "   - ");
        if (qp->userJobLimit < INFINIT_INT)
            sprintf (userJobLimit, "%4d", qp->userJobLimit);
        else
            sprintf (userJobLimit, "  - ");
        if (qp->procJobLimit < INFINIT_FLOAT)
            sprintf (procJobLimit, "%4.1f", qp->procJobLimit);
        else
            sprintf (procJobLimit, "  - ");
        if (qp->hostJobLimit < INFINIT_INT)
            sprintf (hostJobLimit, "%4d", qp->hostJobLimit);
        else
            sprintf (hostJobLimit, "  - ");

        if (i > 0)
            printf("-------------------------------------------------------------------------------\n\n");
        printf("QUEUE: %s\n", qp->queue);
        printf("  -- %s", qp->description);
        if (qp->qAttrib & Q_ATTRIB_DEFAULT) {
            if (numDefaults == 1)
                printf("  This is the default queue.\n\n");
            else
                printf("  This is one of the default queues.\n\n");
        } else
            printf("\n\n");

        printf("PARAMETERS/STATISTICS\n");
        printf("PRIO");
        printf(" NICE");
        printf("     STATUS       ");
        printf("MAX JL/U JL/P JL/H ");
        printf("NJOBS  PEND  RUN  SSUSP USUSP  RSV\n");
        printf("%4d", qp->priority);
        printf(" %3d", qp->nice);
        printf(" %-16.16s", status);

        printf("%s %s %s %s",
               maxJobs, userJobLimit, procJobLimit,hostJobLimit);

        printf("%5d %5d %5d %5d %5d %4d\n\n",
               qp->numJobs, qp->numPEND, qp->numRUN,
               qp->numSSUSP, qp->numUSUSP, qp->numRESERVE);

        if (qp->mig < INFINIT_INT)
            printf("Migration threshold is %d minutes\n", qp->mig);

        if (qp->schedDelay < INFINIT_INT)
            printf("Schedule delay for a new job is %d seconds\n",
                   qp->schedDelay);

        if (qp->acceptIntvl < INFINIT_INT)
            printf("Interval for a host to accept two jobs is %d seconds\n",
                   qp->acceptIntvl);


        if (((qp->defLimits[LSF_RLIMIT_CPU] != INFINIT_INT) &&
             (qp->defLimits[LSF_RLIMIT_CPU] > 0 )) ||
            ((qp->defLimits[LSF_RLIMIT_RUN] != INFINIT_INT) &&
             (qp->defLimits[LSF_RLIMIT_RUN] > 0 )) ||
            ((qp->defLimits[LSF_RLIMIT_DATA] != INFINIT_INT) &&
             (qp->defLimits[LSF_RLIMIT_DATA] > 0 )) ||
            ((qp->defLimits[LSF_RLIMIT_RSS] != INFINIT_INT) &&
             (qp->defLimits[LSF_RLIMIT_RSS] > 0 )) ||
            ((qp->defLimits[LSF_RLIMIT_PROCESS] != INFINIT_INT) &&
             (qp->defLimits[LSF_RLIMIT_PROCESS] > 0 ))) {


            printf("\n");
            printf("DEFAULT LIMITS:");
            prtResourceLimit (qp->defLimits, qp->hostSpec, 1.0, 0);
            printf("\n");
            printf("MAXIMUM LIMITS:");
        }

        procLimits[0] = qp->minProcLimit;
        procLimits[1] = qp->defProcLimit;
        procLimits[2] = qp->procLimit;
        prtResourceLimit (qp->rLimits, qp->hostSpec, 1.0, procLimits);

        printf("\nSCHEDULING PARAMETERS\n");
        if (printThresholds(qp->loadSched,  qp->loadStop, NULL, NULL,
                            MIN(lsInfo->numIndx, qp->nIdx), lsInfo) < 0)
            exit(-1);

        if ((qp->qAttrib & Q_ATTRIB_EXCLUSIVE)
            || (qp->qAttrib & Q_ATTRIB_BACKFILL)
            || (qp->qAttrib & Q_ATTRIB_IGNORE_DEADLINE)
            || (qp->qAttrib & Q_ATTRIB_ONLY_INTERACTIVE)
            || (qp->qAttrib & Q_ATTRIB_NO_INTERACTIVE)
            || (qp->qAttrib & Q_ATTRIB_FAIRSHARE)) {
            printf("\nSCHEDULING POLICIES:");
            if (qp->qAttrib & Q_ATTRIB_FAIRSHARE)
                printf("  FAIRSHARE");
            if (qp->qAttrib & Q_ATTRIB_BACKFILL)
                printf("  BACKFILL");
            if (qp->qAttrib & Q_ATTRIB_IGNORE_DEADLINE)
                printf("  IGNORE_DEADLINE");
            if (qp->qAttrib & Q_ATTRIB_EXCLUSIVE)
                printf("  EXCLUSIVE");
            if (qp->qAttrib & Q_ATTRIB_NO_INTERACTIVE)
                printf("  NO_INTERACTIVE");
            if (qp->qAttrib & Q_ATTRIB_ONLY_INTERACTIVE)
                printf("  ONLY_INTERACTIVE");
	    if (qp->qAttrib & Q_ATTRIB_OWNERSHIP)
		printf("  OWNED_SLOTS");
            if (qp->qAttrib & Q_ATTRIB_ROUND_ROBIN)
                printf("ROUND_ROBIN_SCHEDULING:  yes\n");
            printf("\n");
        }

        /* If the queue has the FAIRSHARE policy on, print out
         * shareAcctInforEnt data structure.
         */
        if (qp->qAttrib & Q_ATTRIB_FAIRSHARE) {
            print_slot_shares(qp);
        }

	if (qp->qAttrib & Q_ATTRIB_OWNERSHIP) {
	    print_slot_owned(qp);
	}

        if (qp->qAttrib & Q_ATTRIB_PREEMPTIVE)
            printf("\nPREEMPTION = %s", qp->preemption);

        if (strcmp (qp->defaultHostSpec, " ") !=  0)
            printf("\nDEFAULT HOST SPECIFICATION:  %s\n", qp->defaultHostSpec);

        if (qp->windows && strcmp (qp->windows, " " ) !=0)
            printf("\nRUN_WINDOWS:  %s\n", qp->windows);
        if (strcmp (qp->windowsD, " ")  !=  0)
            printf("\nDISPATCH_WINDOW:  %s\n", qp->windowsD);

        if ( strcmp(qp->userList, " ") == 0) {
            printf("\nUSERS:  all users\n");
        } else {
            if (strcmp(qp->userList, " ") != 0 && qp->userList[0] != 0)
                printf("\nUSERS:  %s\n", qp->userList);
        }

        if (strcmp(qp->hostList, " ") == 0) {
            printf("HOSTS:  all hosts used by the OpenLava system\n");
        } else {
            if (strcmp(qp->hostList, " ") != 0 && qp->hostList[0])
                printf("HOSTS:  %s\n", qp->hostList);
        }
        if (strcmp (qp->admins, " ") != 0)
            printf("ADMINISTRATORS:  %s\n", qp->admins);
        if (strcmp (qp->preCmd, " ") != 0)
            printf("PRE_EXEC:  %s\n", qp->preCmd);
        if (strcmp (qp->postCmd, " ") != 0)
            printf("POST_EXEC:  %s\n", qp->postCmd);
        if (strcmp (qp->requeueEValues, " ") != 0)
            printf("REQUEUE_EXIT_VALUES:  %s\n", qp->requeueEValues);
        if (strcmp (qp->resReq, " ") != 0)
            printf("RES_REQ:  %s\n", qp->resReq);
        if (qp->slotHoldTime > 0
	    && !(qp->qAttrib & Q_ATTRIB_MEM_RESERVE))
            printf("Maximum slot reservation time: %d seconds\n", qp->slotHoldTime);
        if (qp->slotHoldTime > 0
	    && (qp->qAttrib & Q_ATTRIB_MEM_RESERVE))
            printf("Maximum slot and memory reservation time: %d seconds\n",
		   qp->slotHoldTime);
        if (strcmp (qp->resumeCond, " ") != 0)
            printf("RESUME_COND:  %s\n", qp->resumeCond);
        if (strcmp (qp->stopCond, " ") != 0)
            printf("STOP_COND:  %s\n", qp->stopCond);
        if (strcmp (qp->jobStarter, " ") != 0)
            printf("JOB_STARTER:  %s\n", qp->jobStarter);

        /* CONF_SIG_ACT */

        printf("\n");
        printFlag = 0;
        if  ((qp->suspendActCmd != NULL)
             && (qp->suspendActCmd[0] != ' '))
            printFlag = 1;

        printFlag1 = 0;
        if  ((qp->resumeActCmd != NULL)
             && (qp->resumeActCmd[0] != ' '))
            printFlag1 = 1;

        printFlag2 = 0;
        if  ((qp->terminateActCmd != NULL)
             && (qp->terminateActCmd[0] != ' '))
            printFlag2 = 1;

        if (printFlag || printFlag1 || printFlag2)
            printf("JOB_CONTROLS:\n");


        if (printFlag) {
            printf("    SUSPEND:  ");
            if (strcmp (qp->suspendActCmd, " ") != 0)
                printf("    [%s]\n", qp->suspendActCmd);
        }

        if (printFlag1) {
            printf("    RESUME:   ");
            if (strcmp (qp->resumeActCmd, " ") != 0)
                printf("    [%s]\n", qp->resumeActCmd);
        }

        if (printFlag2) {
            printf("    TERMINATE:");
            if (strcmp (qp->terminateActCmd, " ") != 0)
                printf("    [%s]\n", qp->terminateActCmd);
        }

        if (printFlag || printFlag1 || printFlag2)
            printf("\n");

        printFlag = terminateWhen_(qp->sigMap, "USER");
        printFlag1 = terminateWhen_(qp->sigMap, "PREEMPT");
        printFlag2 = terminateWhen_(qp->sigMap, "WINDOW");
        printFlag3 = terminateWhen_(qp->sigMap, "LOAD");

        if (printFlag | printFlag1 | printFlag2 | printFlag3) {
            printf("TERMINATE_WHEN = ");
            if (printFlag) printf("USER ");
            if (printFlag1) printf("PREEMPT ");
            if (printFlag2) printf("WINDOW ");
            if (printFlag3) printf("LOAD");
            printf("\n");
        }
    }
}

/* prtQueuesSort()
 */
static void
prtQueuesShort(int numQueues, struct queueInfoEnt *queueInfo)
{
    struct queueInfoEnt *qp;
    char *status;
    char first = FALSE;
    int i;
    char userJobLimit[MAX_CHARLEN];
    char procJobLimit[MAX_CHARLEN];
    char hostJobLimit[MAX_CHARLEN];
    char maxJobs[MAX_CHARLEN];

    if(!first) {
        printf("QUEUE_NAME     PRIO      STATUS      ");

        printf("MAX  JL/U JL/P JL/H ");

        printf("NJOBS  PEND  RUN  SUSP\n");
        first = TRUE;
    }
    for (i = 0; i < numQueues; i++) {

        qp = &(queueInfo[i]);

        if (qp->qStatus & QUEUE_STAT_OPEN) {
            if ((qp->qStatus & QUEUE_STAT_ACTIVE)
                && (qp->qStatus & QUEUE_STAT_RUN))
                status = "  Open:Active  ";
            if ((qp->qStatus & QUEUE_STAT_ACTIVE)
                && !(qp->qStatus & QUEUE_STAT_RUN))
                status = " Open:Inactive ";
            if (!(qp->qStatus & QUEUE_STAT_ACTIVE))
                status = " Open:Inactive "; /* By user (manager) */
        } else {  /* queue is closed */
            if ((qp->qStatus & QUEUE_STAT_ACTIVE)
                && (qp->qStatus & QUEUE_STAT_RUN))
                status = " Closed:Active ";
            if ((qp->qStatus & QUEUE_STAT_ACTIVE)
                && !(qp->qStatus & QUEUE_STAT_RUN))
                status = "Closed:Inactive"; /* by time window */
            if (!(qp->qStatus & QUEUE_STAT_ACTIVE))
                status = "Closed:Inactive"; /* by user (manager) */
        }
        if (qp->maxJobs < INFINIT_INT)
            sprintf (maxJobs, "%5d", qp->maxJobs);
        else
            sprintf (maxJobs, "    -");
        if (qp->userJobLimit < INFINIT_INT)
            sprintf (userJobLimit, "%4d", qp->userJobLimit);
        else
            sprintf (userJobLimit, "   -");
        if (qp->procJobLimit < INFINIT_FLOAT)
            sprintf (procJobLimit, "%4.0f", qp->procJobLimit);
        else
            sprintf (procJobLimit, "   -");
        if (qp->hostJobLimit < INFINIT_INT)
            sprintf (hostJobLimit, "%4d", qp->hostJobLimit);
        else
            sprintf (hostJobLimit, "   -");

        if (wflag) {
            printf("%-15s %2d  %-15s", qp->queue, qp->priority, status);
            printf("%s %s %s %s", maxJobs,
                   userJobLimit, procJobLimit, hostJobLimit);

            printf(" %5d %5d %5d %5d \n",
                   qp->numJobs, qp->numPEND, qp->numRUN,
                   qp->numSSUSP + qp->numUSUSP);
        } else {

            printf("%-15.15s %2d  %-15.15s",
                   qp->queue, qp->priority, status);

            printf("%5s %4s %4s %4s", maxJobs,
                   userJobLimit, procJobLimit, hostJobLimit);

            printf(" %5d %5d %5d %5d \n",
                   qp->numJobs, qp->numPEND, qp->numRUN,
                   qp->numSSUSP + qp->numUSUSP);
        }
    }
}

/* print_slot_shares()
 */
static void
print_slot_shares(struct queueInfoEnt *qp)
{
    int i;
    uint32_t numRUN;

    numRUN = 0;
    for (i = 0; i < qp->numAccts; i++) {
        if (qp->saccts[i]->options & SACCT_USER)
            numRUN = numRUN + qp->saccts[i]->numRUN;
    }

    printf("\
\nTOTAL_SLOTS: %u FREE_SLOTS: %u\n", qp->numFairSlots,
           qp->numFairSlots - numRUN);

    printf("\
%9s   %6s   %8s   %6s   %6s\n",
           "USER/GROUP", "SHARES", "PRIORITY", "PEND", "RUN");
    for (i = 0; i < qp->numAccts; i++) {
        char buf[MAXLSFNAMELEN];

        if (qp->saccts[i]->options & SACCT_GROUP)
            sprintf(buf, "%s/", qp->saccts[i]->name);
        else
            sprintf(buf, "%s", qp->saccts[i]->name);

        printf("%-9s   %6d    %8.3f",
               buf,
               qp->saccts[i]->shares,
               qp->saccts[i]->dshares);
        printf("   %6d   %6d\n",
               qp->saccts[i]->numPEND,
               qp->saccts[i]->numRUN);
    }
}

/* print_slot_owned()
 */
static void
print_slot_owned(struct queueInfoEnt *qp)
{
    int i;
    uint32_t numRUN;

    numRUN = 0;
    for (i = 0; i < qp->numAccts; i++) {
        if (qp->saccts[i]->options & SACCT_USER)
            numRUN = numRUN + qp->saccts[i]->numRUN;
    }

    printf("\
\nTOTAL_SLOTS: %u FREE_SLOTS: %u\n", qp->num_owned_slots,
           qp->num_owned_slots - numRUN);

    printf("\
%9s   %6s   %6s   %6s   %8s\n",
           "USER/GROUP", "OWNED", "PEND", "RUN", "BORROWED");
    for (i = 0; i < qp->numAccts; i++) {
        char buf[MAXLSFNAMELEN];

        if (qp->saccts[i]->options & SACCT_GROUP)
            sprintf(buf, "%s/", qp->saccts[i]->name);
        else
            sprintf(buf, "%s", qp->saccts[i]->name);

        printf("%-9s   %6d",
               buf,
               qp->saccts[i]->shares);
        printf("   %6d   %6d   %8d\n",
               qp->saccts[i]->numPEND,
               qp->saccts[i]->numRUN,
	       qp->saccts[i]->numBORROWED);
    }
}
